promql/info.go

```diff
@@ -0,0 +1,464 @@
+// Copyright 2024 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package promql
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/grafana/regexp"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/util/annotations"
+)
+
+const targetInfo = "target_info"
+
+// identifyingLabels are the labels we consider as identifying for info metrics.
+// Currently hard coded, so we don't need knowledge of individual info metrics.
+var identifyingLabels = []string{"instance", "job"}
+
+// evalInfo implements the info PromQL function.
+func (ev *evaluator) evalInfo(ctx context.Context, args parser.Expressions) (parser.Value, annotations.Annotations) {
+	val, annots := ev.eval(ctx, args[0])
+	mat := val.(Matrix)
+	// Map from data label name to matchers.
+	dataLabelMatchers := map[string][]*labels.Matcher{}
+	var infoNameMatchers []*labels.Matcher
+	if len(args) > 1 {
+		// TODO: Introduce a dedicated LabelSelector type.
+		labelSelector := args[1].(*parser.VectorSelector)
+		for _, m := range labelSelector.LabelMatchers {
+			dataLabelMatchers[m.Name] = append(dataLabelMatchers[m.Name], m)
+			if m.Name == labels.MetricName {
+				infoNameMatchers = append(infoNameMatchers, m)
+			}
+		}
+	} else {
+		infoNameMatchers = []*labels.Matcher{labels.MustNewMatcher(labels.MatchEqual, labels.MetricName, targetInfo)}
+	}
+
+	// Don't try to enrich info series.
+	ignoreSeries := map[int]struct{}{}
+loop:
+	for i, s := range mat {
+		lblMap := s.Metric.Map()
+		name := lblMap[labels.MetricName]
+		for _, m := range infoNameMatchers {
+			if m.Matches(name) {
+				ignoreSeries[i] = struct{}{}
```

Comment:
I'm a bit lost here.  We are ignoring certain rows from the LHS depending on the name, if the name is a match on the RHS?
Do you have an example where this is necessary?

Reply:
It's as the comment before the loop says: "Don't try to enrich info series.". `ignoreSeries` is supposed to contain indices of (LHS) series that are info series (given they have names of known info metrics). The series contained in `ignoreSeries` are ignored when it comes to picking labels from info series, since they are themselves info series. Make sense?

An example (taken from the `info` tests) is the query `info(build_info, {__name__=~".+_info", build_data=~".+"})`. If the block in question is removed, the test fails due to `__name__=~".+_info"` matching `target_info`, which misses the requested `build_data` label. As per `info` semantics, no `build_info` series are then returned (as there are no matching info series).

Reply:
Your example actually joins `build_info` with nothing?  Is that a good example?

Do we expect the LHS to consist of several different metrics?

Reply:
It's not intended as a good example, but rather a corner case that a user might attempt. Some info metrics also don't have the "_info" suffix, and in the future `info` might recognize info metrics automatically (through persisted metadata). Therefore, calling `info` with an info metric as LHS is worth guarding against I think.

> Do we expect the LHS to consist of several different metrics?

It's certainly possible to produce such a query.